### ISSUE TITLE
api/compare: distinguish the "best" benchmark for a given case/context/run

### ISSUE THREAD

**Issue by austin3dickey** [conbench/conbench#955]
In [some use cases](https://github.com/facebookincubator/velox/blob/89ee8a127455676984b68e5037ce295cf68848c6/.github/workflows/benchmark.yml#L158), as a way of addressing #572, we are sending multiple BenchmarkResults with the same case and context to the same Run. If the first invocation of the benchmark resulted in a suspected regression (via client-side logic until #635), they run the same benchmark again in order to see if the first invocation was just an outlier. They then post this result to the same Run.

This breaks a few assumptions in our codebase. The `/api/compare` endpoint [deduplicates](https://github.com/conbench/conbench/blob/2ad32a8bb0505d41941b0d8e6ee8e5390c868d53/conbench/api/compare.py#L67) these results in (IMO) an undefined manner. The `benchalerts` library hits this endpoint to determine whether each result displays a lookback z-score regression, but since it is using an undefined result, it's not necessarily making this decision in the way the user might want to make the decision.

It would be useful for the compare API endpoints to distinguish which result is the "best" of the results in one of those groups, because the best result is the one that users would care about for alerting purposes.

"best" would be determined by: which result's `min` is the lowest (for `_less_is_better(unit)=True`) or `max` is the highest (for `_less_is_better(unit)=False`)?

**Comment by alistaire47** [comment]
We'll need to deal with this situation for automatic retries as we get there too; as I understand, the workflow will be roughly

1. run a benchmark
2. submit the result
3. get a response that says it's a regression or outlier
4. rerun the benchmark
5. resubmit result

...so we should be able to sensibly handle at least the cases where the first run is an outlier and where there is a mean shift (so the second result is similar to the first). (There's some uncertainty here—you could get two outliers in a row or other corner cases—but handling the basic cases gets us pretty far.)

I'm not sure everywhere we'll need to handle these—the compare endpoint as you say, and plots somehow—but generally I think the move here is to sort by benchmark run time instead of commit time.

**Comment by austin3dickey** [comment]
>The /api/compare endpoint [deduplicates](https://github.com/conbench/conbench/blob/2ad32a8bb0505d41941b0d8e6ee8e5390c868d53/conbench/api/compare.py#L67) these results in (IMO) an undefined manner.

I am going to tackle this work in this ticket, as it's a prerequisite.

**Comment by jonkeane** [comment]
> I am going to tackle this work in this ticket, as it's a prerequisite.

Did you mean to link to a different ticket here?

**Comment by austin3dickey** [comment]
Nope, I meant that this ticket should include the compare API returning all results, instead of deduplicating, and then label one of them as the "best" one to use for alerting. Does that make sense to everyone else?

**Comment by jonkeane** [comment]
Ah ok, got it. 

I suspect figuring out "best" here is going to be a can of worms at this point (though once we can do things like using aggregations other than mean across iterations it will become much easier — at a minimum there would be a framework to slot this into). I wonder if for now we do what we've been doing with iterations and take the mean across all of the results or treat each result as if it was an independent thing (for history purposes this is probably the easiest, though what we call the default baseline is tricky — maybe there we say the most recent result wins any ties).

**Comment by austin3dickey** [comment]
Oof, I see the source of some confusion - this ticket was originally talking about the `/api/benchmarks` endpoint and I forgot to change it to the `/api/compare` endpoints. Let me do that now.

I originally opened this issue to solve the use case of "a user is retrying benchmarks and sending multiple results to one run that share case/context - for a given case/context, how do we choose whether to alert or not?" When we talked earlier, we agreed to "pick one and go with it," and I decided to open this ticket to propose that the one we choose is 

>which result's min is the lowest (for _less_is_better(unit)=True) or max is the highest (for _less_is_better(unit)=False)

But we could aggregate by `mean` instead, or choose the one with the latest timestamp.

In framing this ticket, I proposed that we should still return _all_ comparisons from the endpoint, and add an extra flag `"is_best": True` to distinguish which one to use for alerting. But maybe since we might aggregate in other ways for other use cases, I should scope the ticket down to something like `"use_for_alerting": True`?

**Comment by jgehrcke** [comment]
> "best" would be determined by: which result's min is the lowest (for _less_is_better(unit)=True) or max is the highest (for _less_is_better(unit)=False)?

Relates to

https://github.com/conbench/conbench/issues/1335
https://github.com/conbench/conbench/issues/535
